RandomSource update - Validated

c19x opened the pull request:
- Enables developers to supply their own custom RandomSource
  - Same four existing sources are still available
  - Updated RandomSource test to ensure test for randomness should always pass
- PseudoDeviceAddress (PDA) now use SHA256 to cryptographically separate random address from random source
  - Existing non-blocking random source made it impractical to use observed addresses for tracking
    - Call time is inherently random due to real-world events as entropy source
    - Shared random seed source is inherently random as system initialisation time is unpredictable
    - Random values follow a uniform distribution, thus all values may be generated by all seeds
    - Random sequence follow a uniform distribution, thus knowledge of prior values offer minimal gain in reducing the total number of seeds that could have generated the consecutive addresses
  - Introduction of SHA256 hash and data trimming offer additional assurance that the random seed is cryptographically challenging to identify given observed addresses
    - Developers are free to adopt their own customer RandomSource that meets their specific requirements
    - Recommendation is to use the non-blocking random source in contact tracing applications as the impact of blocking, leading to non-detection of disease transmission, is far greater than the theoretical risk of tracking
- Passes all unit tests

Signed-off-by: c19x <support@c19x.org>

closes #189

adamfowleruk commented on herald/src/test/java/io/heraldprox/herald/sensor/datatype/RandomSourceTests.java:
```diff
@@ -7,63 +7,85 @@
                 lastValue = value;
             }
+            // Duplicates may occur, but should be unlikely
+            assertTrue(duplicates < 900);
```
90% seems rather high for a tolerance.

c19x replied:
Updated comment to highlight this is for catching obvious errors and added uniformity test.

adamfowleruk commented on herald/src/test/java/io/heraldprox/herald/sensor/datatype/RandomSourceTests.java:
```diff
@@ -7,63 +7,85 @@
                 lastValue = value;
             }
+            // Duplicates may occur, but should be unlikely
+            assertTrue(duplicates < 900);
```
90% seems rather high for a tolerance.

c19x replied:
Updated comment to highlight this is for catching obvious errors and added uniformity test.

adamfowleruk commented on herald/src/test/java/io/heraldprox/herald/sensor/datatype/RandomSourceTests.java:
```diff
@@ -7,63 +7,85 @@
                 lastValue = value;
             }
+            // Duplicates may occur, but should be unlikely
+            assertTrue(duplicates < 900);
```
90% seems rather high for a tolerance.

c19x replied:
Updated comment to highlight this is for catching obvious errors and added uniformity test.

adamfowleruk commented:
Would it be possible to add a higher level test for PseudoMacAddress that validated each byte value for the mac had an even distribution and was comparable in every byte position? (I.e. to ensure that the first byte being, say, 00 was just as likely as its other values, and that the frequency was comparable with bytes 2-6's frequency of 00 too).

c19x commented:
Thanks for review.
- Current tests have a high tolerance as they aim to test for obviously not random output (i.e. due to coding error), rather than confirm the output is random and uniformly distributed
- Good idea to incorporate uniformity tests, considering incorporating existing standalone simulation code, but need to make it run within reasonable time
- Also have an idea for non-blocking secure random, shall provide update, please hold PR

c19x commented:
- Added comment to clarify why the tolerance is so high for existing tests (to catch obviously not random output)
- Added new uniformity tests with just 10% threshold for confirming if all values are equally likely in a sequence and as initial value given a seed
  - Uniformity test can run at lower threshold given more samples, current parameters have been set to ensure the automated tests should run quickly and passes almost all the time given uniformly distributed random sources
- Introduced new Non-blocking CSPRNG based on the same concepts as SecureRandom except entropy is gathered from sources that won't exhaust in the target application environment. This is a simplified and quantified version of existing Non-blocking PRNG, enhanced with SHA256 for separating the observations (random values) generated by the one-time use PRNG from its seed that has been derived from a mixture of entropy sources, and also separating the seed source from the entropy source.
  - Passes all new unit tests
  - Delivers the best speed performance across all available random sources.
  - This is now the recommended default random source.
  - Validated on Pixel 3 and A20
- Running overnight test to confirm new Non-Blocking CSPRNG really is non-blocking on idle devices
  - Code analysis has not identified any blocking calls in the new random source.
  - Planning to migrate uniformity test to server for statistical tests as separate standalone background task.

c19x commented:
- Conducted 18hr+ test on Pixel3 and A20 to confirm Non-Blocking CSPRNG is non-blocking

c19x commented:
- NonBlockingCSPRNG confirmed to be non-blocking after 44hr test on idle devices
- Battery usage was 84% and 82% after 44 hours continuous use
- Added new NonBlockingSecureRandom which is the recommended random source for the future as all reliance on LCG has been removed. The solution is based entirely on cryptographically secure functions only, yet remains non-blocking.
- Conducting long running test to confirm NonBlockingSecureRandom now.

c19x commented:
- NonBlockingSecureRandom **FAILED** to perform continuously without any negative impact on wider systems
  - Test shows any call to SecureRandom is liable to exhaust system entropy which prevents BLE operation
  - Developing new algorithm to remove reliance on SecureRandom

c19x commented:
- New implementation of NonBlockingSecureRandom is guaranteed to be non-blocking
  - Eliminated reliance on SecureRandom for reseeding
  - Using a time-based truly unpredictable random source for reseeding
    - Solution is based on thread scheduling which is inherently unpredictable and determined by wider system events
    - Same concept as ThreadSeedGenerator that is used by Oracle and OpenJDK implementations of SecureRandom that acts as backup random source when /dev/random is exhausted
- Running overnight test under stress testing condition to confirm solution is non-blocking
  - No logging to avoid generating storage events to minimise system entropy
  - Advert refreshes every 3 minutes to trigger generation of new pseudo device address
  - RandomSource gathering external entropy from just one target device (test conducted with 2 phones in a faraday bag)
  - NonBlockingSecureRandom set to reseed every minute

c19x commented:
- Tests have confirmed NonBlockingSecureRandom is non-blocking under continuous background use
  - Advert refresh set to once every 3 minutes to trigger unusually frequent use of random source
  - Reseed set to once every minute to trigger unusually frequent use of entropy source
  - Logging set to OFF to minimise system activities
  - Pixel3 and A20 placed in a single Faraday bag to eliminate external entropy
  - 100% continuity in 5 hour test, shall continue testing for 24 hours
- PR ready for merge as new solution is now validated and proven to be non-blocking
  - Design builds on accepted seed generation method in SecureRandom and addresses known vulnerabilities
  - 2048 bit internal state, SHA256 based state updates, incorporates external truly random data when available

closes #189 
closes #118 
closes #110

c19x commented:
Ready for merge
- 27hr test shows 100% continuity
